### PR TITLE
Refactor: track MDC scope entries per-thread

### LINKED ISSUE #7563: Scope.removeMdcObjectWhenScopeExits seems to have a multi threading issue
### Search first

- [x] I searched and no similar issues were found

### Description

In the junit test suite we have database tests that perform liquibase migrations that are run in parallel, against several different docker databases. Once in a while I get an exception like this:
```
java.lang.ArrayIndexOutOfBoundsException: Index 2 out of bounds for length 0
	at java.base/java.util.ArrayList.add(ArrayList.java:485)
	at java.base/java.util.ArrayList.add(ArrayList.java:497)
	at liquibase.Scope.removeMdcObjectWhenScopeExits(Scope.java:463)
	at liquibase.Scope.addMdcValue(Scope.java:453)
	at liquibase.Scope.addMdcValue(Scope.java:442)
	at liquibase.command.CommandScope.execute(CommandScope.java:233)
```
When I look in the code I can see that this snippet:
```
      addedMdcEntries
                    .computeIfAbsent(currentScope.scopeId, k -> new ArrayList<>())
                    .add(mdcObject);
```
Is not totally thread safe in the sense that the ArrayList.add is not thread safe. We can get into this case when the scopeId is the same. 

The simple fix could be not creating an array list but a concurrent queue or so, but I am not sure this would not break logic down the line as I do not know the internals of liquibase well enough. I am also not sure why the scope id would be the same here, whether that's the objective or whether that's a problem in and of itself.

Liquibase 5.0.1, java 25, Junit 5.11.3, surefire 3.5.4.

### Steps To Reproduce

Difficult to reproduce from the outside, I suspect launching several `removeMdcObjectWhenScopeExits` in the same scope from a junit test should trigger this behaviour down the line.

### Expected/Desired Behavior

I can run my database tests in parallel, including liquibase migrations.

### Liquibase Version

5.0.1

### Database Vendor & Version

postgres 15-alpine docker

### Liquibase Integration

maven

### Liquibase Extensions

_No response_

### OS and/or Infrastructure Type/Provider

reproduced on linux and mac

### Additional Context

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

### LINKED ISSUE #7823: Design: MDC cleanup race across inherited scopes (InheritableThreadLocal) in Scope.java
## Summary

While reviewing PR #7822 (which fixes an `ArrayIndexOutOfBoundsException` in `Scope.removeMdcObjectWhenScopeExits` by switching `addedMdcEntries` to use `CopyOnWriteArrayList`), a related but out-of-scope design concern was identified around MDC cleanup and scope inheritance across threads.

## Background

`Scope` uses an `InheritableThreadLocal<ScopeManager>` so that child threads can inherit the parent's current scope (see `Scope.getScopeManagerThreadLocal()` / `childValue`). Combined with the static `addedMdcEntries` map keyed by `scopeId`, this creates a design issue:

- `Scope.exit(scopeId)` removes and closes all MDC entries tracked under that `scopeId` in `addedMdcEntries`.
- However, if a `scopeId` is shared/inherited across multiple threads (via `InheritableThreadLocal`), another thread may still be actively using that same scope and calling `addMdcValue(...)` *after* the entries were already cleaned up by `exit()` on a different thread.
- Such late `addMdcValue()` calls would populate a *new* list in `addedMdcEntries` for that `scopeId` (since it was just removed), which won't be cleared until/unless that same scope exits again on some thread — potentially leaving MDC state hanging longer than expected, or being cleared at an unexpected time.

This is a structural/architectural concern regarding how `scopeId`s are shared and cleaned up across inherited child threads, distinct from the thread-safety fix in #7822 (which only addresses the `ArrayIndexOutOfBoundsException` from concurrent `.add()` calls on a non-thread-safe `ArrayList`).

## Affected area

- `liquibase-standard/src/main/java/liquibase/Scope.java`
  - `getScopeManagerThreadLocal()` (the `InheritableThreadLocal` and its `childValue` override)
  - `exit(String scopeId)`
  - `removeMdcObjectWhenScopeExits(...)`
  - `addedMdcEntries` static map

## Suggested direction

- Reconsider whether scope/MDC entry ownership should be tracked per-thread rather than purely per-`scopeId`, or
- Add explicit coordination (e.g., reference counting per scope across threads) so that `exit()` on one thread doesn't prematurely tear down MDC state still in use by another thread sharing the same inherited scope, or
- Add a regression test that exercises: parent thread creates a scope, spawns a child thread that inherits it via `InheritableThreadLocal`, child thread calls `addMdcValue`, parent thread calls `exit()`, and verify expected MDC cleanup behavior.

## References

- Related issue: #7563
- PR: https://github.com/liquibase/liquibase/pull/7822
- Discussion: https://github.com/liquibase/liquibase/pull/7822#discussion_r3501425888

## Acceptance criteria

- [ ] Design decision documented on how MDC entry ownership/cleanup should behave when a scope is inherited across threads.
- [ ] Implementation (if needed) to prevent premature or delayed MDC cleanup in this scenario.
- [ ] Regression test covering inherited-scope MDC add/exit ordering across threads.

### LINKED ISSUE #7824: Design: MDC cleanup race across inherited scopes (InheritableThreadLocal) in Scope.java
## Summary

While reviewing PR #7822 (which fixes an `ArrayIndexOutOfBoundsException` in `Scope.removeMdcObjectWhenScopeExits` by switching `addedMdcEntries` to use `CopyOnWriteArrayList`), a related but out-of-scope design concern was identified around MDC cleanup and scope inheritance across threads.

## Background

`Scope` uses an `InheritableThreadLocal<ScopeManager>` so that child threads can inherit the parent's current scope (see `Scope.getScopeManagerThreadLocal()` / `childValue`). Combined with the static `addedMdcEntries` map keyed by `scopeId`, this creates a design issue:

- `Scope.exit(scopeId)` removes and closes all MDC entries tracked under that `scopeId` in `addedMdcEntries`.
- However, if a `scopeId` is shared/inherited across multiple threads (via `InheritableThreadLocal`), another thread may still be actively using that same scope and calling `addMdcValue(...)` *after* the entries were already cleaned up by `exit()` on a different thread.
- Such late `addMdcValue()` calls would populate a *new* list in `addedMdcEntries` for that `scopeId` (since it was just removed), which won't be cleared until/unless that same scope exits again on some thread — potentially leaving MDC state hanging longer than expected, or being cleared at an unexpected time.

This is a structural/architectural concern regarding how `scopeId`s are shared and cleaned up across inherited child threads, distinct from the thread-safety fix in #7822 (which only addresses the `ArrayIndexOutOfBoundsException` from concurrent `.add()` calls on a non-thread-safe `ArrayList`).

## Affected area

- `liquibase-standard/src/main/java/liquibase/Scope.java`
  - `getScopeManagerThreadLocal()` (the `InheritableThreadLocal` and its `childValue` override)
  - `exit(String scopeId)`
  - `removeMdcObjectWhenScopeExits(...)`
  - `addedMdcEntries` static map

## Suggested direction

- Reconsider whether scope/MDC entry ownership should be tracked per-thread rather than purely per-`scopeId`, or
- Add explicit coordination (e.g., reference counting per scope across threads) so that `exit()` on one thread doesn't prematurely tear down MDC state still in use by another thread sharing the same inherited scope, or
- Add a regression test that exercises: parent thread creates a scope, spawns a child thread that inherits it via `InheritableThreadLocal`, child thread calls `addMdcValue`, parent thread calls `exit()`, and verify expected MDC cleanup behavior.

## References

- Related issue: #7563
- PR: https://github.com/liquibase/liquibase/pull/7822
- Discussion: https://github.com/liquibase/liquibase/pull/7822#discussion_r3501425888
- Requested by: @om7057

## Acceptance criteria

- [ ] Design decision documented on how MDC entry ownership/cleanup should behave when a scope is inherited across threads.
- [ ] Implementation (if needed) to prevent premature or delayed MDC cleanup in this scenario.
- [ ] Regression test covering inherited-scope MDC add/exit ordering across threads.

### PR DESCRIPTION
**Pull request by om7057**
## Description
This PR resolves #[7823](https://github.com/liquibase/liquibase/issues/7823) and serves as an architectural follow-up to PR #7822.

Previously, `addedMdcEntries` was a global static map keyed purely by `scopeId`. Because `Scope` is inherited across threads via `InheritableThreadLocal`, a child thread could add MDC values to an inherited scope, only for the parent thread's `exit()` call to prematurely wipe out the global tracking list. This left the child thread's MDC state permanently hanging.

## Changes Made
- Wrapped `addedMdcEntries` in a `ThreadLocal<Map<String, List<MdcObject>>>`.
- Because the map is now isolated per-thread, concurrent access is impossible. We safely reverted `ConcurrentHashMap` and `CopyOnWriteArrayList` back to standard `HashMap` and `ArrayList`.
- Now, when a scope is inherited, each thread independently tracks and cleans up its own MDC additions upon exiting the scope.

## Testing
- Verified successful compilation.
- Passed all unit tests in `liquibase-standard`.

Signed-off-by: om7057 <kulkarniom7057@gmail.com>

Fixes #7823
Fixes #7824 (duplicate of #7823)
Fixes #7563

### CODE REVIEW

**Comment by om7057** [comment]
I've updated the implementation to lazily initialize the ThreadLocal map only when an MDC entry is added, and it now explicitly calls ThreadLocal.remove() when the map is emptied upon scope exit. This prevents retaining empty HashMaps on long-lived threads as requested!

File: liquibase-standard/src/main/java/liquibase/Scope.java [hunk 1]
```diff
@@ -291,9 +290,15 @@ public static void exit(String scopeId) throws Exception {
...
+        Map<String, List<MdcObject>> map = addedMdcEntries.get();
+        if (map != null) {
+            List<MdcObject> mdcObjects = map.remove(currentScope.scopeId);
+            for (MdcObject mdcObject : CollectionUtil.createIfNull(mdcObjects)) {
+                mdcObject.close();
+            }
+            if (map.isEmpty()) {
+                addedMdcEntries.remove();
+            }
         }
```
[the remark this answers is not in the document]
Comment (om7057) [reply liquibase-standard/src/main/java/liquibase/Scope.java:326]: Thanks for raising this. After looking into it, I realized that MDC cleanup is fundamentally tied to the thread that writes the value - it's a limitation of Java's ThreadLocal design. A parent thread can't clean up a child thread's MDC context.

The correct pattern is for child threads to wrap their work in `Scope.child(...)`, which ensures cleanup happens on that thread. I've added a regression test in ScopeTest.groovy to demonstrate and verify this behavior.

File: liquibase-standard/src/test/groovy/liquibase/ScopeTest.groovy [hunk 2]
```diff
@@ -231,6 +231,37 @@ class ScopeTest extends Specification {
...
+        Scope.exit(scopeId)
+
+        then:
+        true
+
+        cleanup:
+        mdcFactory.unregister(testManager)
+        mdcFactory.register(existingManager)
+    }
+
```
[the remark this answers is not in the document]
Comment (om7057) [reply liquibase-standard/src/test/groovy/liquibase/ScopeTest.groovy:301]: Great point! You're absolutely right - assertions in child threads would fail silently. I've updated the test to use the error-collection pattern (same as the concurrent scope test) so failures are properly captured and reported. Fixed in commit a5d7e7a.

**Comment by om7057** [comment]
> * [ ]   ▶️ Resume reviews
> *

@coderabbitai resume

**Comment by om7057** [comment]
Hey @filipelautert @wwillard7800: this PR addresses a real bug with MDC cleanup leaking across inherited scopes. Would appreciate a review when you get a chance!

**Review by filipelautert — COMMENTED** [review]
Per-thread tracking matches how MDC actually works, I like this direction better than #7822. Please drop the unrelated /liquibase/ .gitignore line and sign the CLA so we can run CI.

**Comment by om7057** [comment]
> Per-thread tracking matches how MDC actually works, I like this direction better than #7822. Please drop the unrelated /liquibase/ .gitignore line and sign the CLA so we can run CI.

@filipelautert Thanks for the review! Glad the per-thread approach makes sense.

That `/liquibase/` entry was just a leftover from a local build artifact and wasn't related to this PR. I've removed it, and `.gitignore` now matches `main` again (commit `bdae89268`).

Regarding the CLA, it looks like CLA Assistant recognizes it as signed, but it's still showing as pending. I'll check what's holding it up so CI can run, and I'll update the thread once it clears.

**Comment by rberezen** [comment]
Reviewed and verified against the actual bug — the fix is structurally correct, but the included regression test doesn't actually cover the failure it's meant to guard. Details below.

**✅ The fix is right.** Making `addedMdcEntries` per-thread cleanly eliminates the reported cross-thread corruption: on `main`, one thread's `exit()` does `addedMdcEntries.remove(scopeId)` and closes MDC entries that *another* thread registered under the same inherited `scopeId`; per-thread maps mean each thread only ever touches its own. Reverting `ConcurrentHashMap`/`CopyOnWriteArrayList` → `HashMap`/`ArrayList` is safe now that each map is single-thread. And the `if (map.isEmpty()) addedMdcEntries.remove()` in `exit()` is good hygiene — a thread that properly brackets its scopes leaves no `ThreadLocal` residue (important for pool threads, cf. #6891/#7877).

**✅ Verified against the real bug.** I reproduced #7823 deterministically: a child thread inherits the parent's scope (same `scopeId`), registers an MDC entry, and the parent exits that scope first while the child is still logically inside it. On `main` the parent's `exit()` prematurely closes the child's entry; on this branch it survives.
```
main:  child's "sharedKey" GONE after parent exit   → test FAILS
PR:    child's "sharedKey" still present            → test PASSES
```

**⚠️ But the PR's own test passes on `main` too — so it doesn't guard the regression.** I confirmed this by running your `"mdc values added by child thread in a child scope are cleaned up"` test against `main`: it's green there. The reason is that it uses `Scope.child([:])`, which creates a *fresh* child `scopeId` on the child thread — so parent and child never share a `scopeId`, and the cross-thread cleanup collision (the whole bug) never happens. The bug specifically requires the **same inherited `scopeId`** used across threads.

Here's a deterministic test that fails on `main` and passes here — worth using in place of (or alongside) the current one:
```groovy
def "parent exit must not close a child thread's MDC entry under a shared inherited scope"() {
    given:
    def mdcFactory = Scope.currentScope.getSingleton(MdcManagerFactory)
    def existingManager = mdcFactory.getMdcManager()
    def testManager = new TestMdcManager()
    mdcFactory.unregister(existingManager); mdcFactory.register(testManager)
    def added = new java.util.concurrent.CountDownLatch(1)
    def parentExited = new java.util.concurrent.CountDownLatch(1)
    def present = new java.util.concurrent.atomic.AtomicBoolean(true)
    def scopeId = Scope.enter(null, [:])          // parent scope; child inherits the same scopeId

    when:
    def child = Thread.start {
        Scope.getCurrentScope().addMdcValue("sharedKey", "v", true)
        added.countDown()
        parentExited.await(30, java.util.concurrent.TimeUnit.SECONDS)
        present.set(testManager.getValues().containsKey("sharedKey"))
    }
    added.await(30, java.util.concurrent.TimeUnit.SECONDS)
    Scope.exit(scopeId)                            // parent exits the shared scope first
    parentExited.countDown(); child.join(30000)

    then:
    present.get()                                  // main: false, PR: true

    cleanup:
    mdcFactory.unregister(testManager); mdcFactory.register(existingManager)
}
```

**Two minor notes:**
- Per-thread cleanup means only the thread that added an entry can clean it. For the normal `Scope.child(enter→exit on one thread)` pattern (which is how commands run) this is exactly correct. The one edge is a thread that *inherits* a scope and adds MDC without its own bracketed child scope+exit — it would never clean that entry (leaks on that thread). Unusual, arguably a usage bug, but worth a comment near `removeMdcObjectWhenScopeExits`.
- Heads-up: #7877 also edits `Scope.java` (the `getCurrentScope`/root-scope region + a new static field near the top). No logical conflict with this PR, but the two will need a clean merge since both add fields near the top of the class.

Solid fix on a genuinely tricky concurrency bug — just needs a test that actually reproduces it. Nice work @om7057.

**Comment by om7057** [comment]
> Reviewed and verified against the actual bug — the fix is structurally correct, but the included regression test doesn't actually cover the failure it's meant to guard. Details below.
> 
> **✅ The fix is right.** Making `addedMdcEntries` per-thread cleanly eliminates the reported cross-thread corruption: on `main`, one thread's `exit()` does `addedMdcEntries.remove(scopeId)` and closes MDC entries that _another_ thread registered under the same inherited `scopeId`; per-thread maps mean each thread only ever touches its own. Reverting `ConcurrentHashMap`/`CopyOnWriteArrayList` → `HashMap`/`ArrayList` is safe now that each map is single-thread. And the `if (map.isEmpty()) addedMdcEntries.remove()` in `exit()` is good hygiene — a thread that properly brackets its scopes leaves no `ThreadLocal` residue (important for pool threads, cf. #6891/#7877).
> 
> **✅ Verified against the real bug.** I reproduced #7823 deterministically: a child thread inherits the parent's scope (same `scopeId`), registers an MDC entry, and the parent exits that scope first while the child is still logically inside it. On `main` the parent's `exit()` prematurely closes the child's entry; on this branch it survives.
> 
> ```
> main:  child's "sharedKey" GONE after parent exit   → test FAILS
> PR:    child's "sharedKey" still present            → test PASSES
> ```
> 
> **⚠️ But the PR's own test passes on `main` too — so it doesn't guard the regression.** I confirmed this by running your `"mdc values added by child thread in a child scope are cleaned up"` test against `main`: it's green there. The reason is that it uses `Scope.child([:])`, which creates a _fresh_ child `scopeId` on the child thread — so parent and child never share a `scopeId`, and the cross-thread cleanup collision (the whole bug) never happens. The bug specifically requires the **same inherited `scopeId`** used across threads.
> 
> Here's a deterministic test that fails on `main` and passes here — worth using in place of (or alongside) the current one:
> 
> ```groovy
> def "parent exit must not close a child thread's MDC entry under a shared inherited scope"() {
>     given:
>     def mdcFactory = Scope.currentScope.getSingleton(MdcManagerFactory)
>     def existingManager = mdcFactory.getMdcManager()
>     def testManager = new TestMdcManager()
>     mdcFactory.unregister(existingManager); mdcFactory.register(testManager)
>     def added = new java.util.concurrent.CountDownLatch(1)
>     def parentExited = new java.util.concurrent.CountDownLatch(1)
>     def present = new java.util.concurrent.atomic.AtomicBoolean(true)
>     def scopeId = Scope.enter(null, [:])          // parent scope; child inherits the same scopeId
> 
>     when:
>     def child = Thread.start {
>         Scope.getCurrentScope().addMdcValue("sharedKey", "v", true)
>         added.countDown()
>         parentExited.await(30, java.util.concurrent.TimeUnit.SECONDS)
>         present.set(testManager.getValues().containsKey("sharedKey"))
>     }
>     added.await(30, java.util.concurrent.TimeUnit.SECONDS)
>     Scope.exit(scopeId)                            // parent exits the shared scope first
>     parentExited.countDown(); child.join(30000)
> 
>     then:
>     present.get()                                  // main: false, PR: true
> 
>     cleanup:
>     mdcFactory.unregister(testManager); mdcFactory.register(existingManager)
> }
> ```
> 
> **Two minor notes:**
> 
> * Per-thread cleanup means only the thread that added an entry can clean it. For the normal `Scope.child(enter→exit on one thread)` pattern (which is how commands run) this is exactly correct. The one edge is a thread that _inherits_ a scope and adds MDC without its own bracketed child scope+exit — it would never clean that entry (leaks on that thread). Unusual, arguably a usage bug, but worth a comment near `removeMdcObjectWhenScopeExits`.
> * Heads-up: [fix: share a single root scope across threads #7877](https://github.com/liquibase/liquibase/pull/7877) also edits `Scope.java` (the `getCurrentScope`/root-scope region + a new static field near the top). No logical conflict with this PR, but the two will need a clean merge since both add fields near the top of the class.
> 
> Solid fix on a genuinely tricky concurrency bug — just needs a test that actually reproduces it. Nice work @om7057.

Thanks for the thorough go-through; you're right, the original test used Scope.child([:]) on the child thread, which mints its own scopeId, so it never actually exercised the shared-scopeId collision. Added your deterministic test verbatim ("parent exit must not close a child thread's MDC entry under a shared inherited scope"), confirmed it fails on main and passes here. Also added a comment near removeMdcObjectWhenScopeExits documenting the inherit-without-bracketing edge case you called out. Noted re: #7877's merge overlap; will keep an eye out.

File: liquibase-standard/src/test/groovy/liquibase/ScopeTest.groovy [hunk 3]
```diff
@@ -231,6 +231,75 @@ class ScopeTest extends Specification {
...
+            added.countDown()
+            parentExited.await(30, java.util.concurrent.TimeUnit.SECONDS)
+            present.set(testManager.getValues().containsKey("sharedKey"))
+        }
+        added.await(30, java.util.concurrent.TimeUnit.SECONDS)
+        Scope.exit(scopeId)                         // parent exits the shared scope first
+        parentExited.countDown(); child.join(30000)
+
+        then:
+        present.get()                               // main: false, PR: true
```
[the remark this answers is not in the document]
Comment (om7057) [reply liquibase-standard/src/test/groovy/liquibase/ScopeTest.groovy:336]: Resolved.

**Comment by om7057** [comment]
@wwillard7800 @filipelautert just a small reminder on this one!

**Review by filipelautert — COMMENTED** [review]
@om7057 the new test does guard it now - ran it here, fails on main and passes on the branch.
CI is stuck on authorize and never ran, a new push should clear it.

**Comment by om7057** [comment]
> @om7057 the new test does guard it now - ran it here, fails on main and passes on the branch. CI is stuck on authorize and never ran, a new push should clear it.

@filipelautert Fresh commit pushed for CI.

**Comment by rberezen** [comment]
Re-verified after the deterministic-test commits (`98218dac6` + `23027becc`) — this now guards the actual bug. 🎉

I ran the new `"parent exit must not close a child thread's MDC entry under a shared inherited scope"` test against a fresh `main` and against this branch:
```
main (static addedMdcEntries):   present.get() == false  → FAILS  (parent's exit wiped the child's entry)
this branch (per-thread):        present.get() == true   → PASSES
```
So unlike the earlier `Scope.child` test, this one genuinely fails on the buggy baseline — real regression coverage. The hardening in `23027becc` is exactly right too: `present` now defaults to `false` and the child's exceptions/liveness are asserted, so a silent no-run can't sneak through as a pass.

Also nice: the doc comment you added on `removeMdcObjectWhenScopeExits` captures the per-thread cleanup contract (only the adding thread cleans up; a thread that inherits a scope and adds MDC without its own enter/exit won't) — good to have that written down.

Full `ScopeTest` is green on the branch (14/14). The fix itself is unchanged and correct. LGTM — this is just waiting on the `authorize`/CI gate now. Thanks for the thorough follow-through @om7057!

**Review by filipelautert — APPROVED** [review]
CI is green now, the only red check is sonar which is broken for every PR at the moment.

**Review by rberezen — APPROVED** [review]
